Lms/add admin snapshot filter options

anathomical opened the pull request:
## WHAT
- Refactor approach used in `options` generation to use PostgreSQL rather than BigQuery for improved stability and user experience
- Add additional options for filters: specifically a list of teachers and a list of classrooms for more specific filtering
- Modify the `options` endpoint to take params of already-selected filter options in order to narrow down higher-specificity filter values based on more general filter selections (e.g. showing only teachers from a selected school)
- Update the `query` endpoint to take additional filter params (`teacher_ids` and `classroom_ids`)
- Update Snapshot cache key generation to handle additional filter params
## WHY
Later versions of the Admin Snapshots project requires these additional filters, and letting the back-end update filter options based on other filter selections should simplify front-end logic.
## HOW
- Write new `OptionsQuery` sub-classes for Teachers and Classrooms
- Add the new queries to the `options` action
- Allow `options` action to accept filter params and pass them through to queries
- Adjust CacheKey generation logic to utilize new params

### Notion Card Links
https://www.notion.so/quill/Backend-for-Admin-Snapshot-Report-a90a9996669a4c60b8a242ce2306cb12?pvs=4

PR Checklist | Your Answer
------------ | -------------
Have you added and/or updated tests? |  Yes
Have you deployed to Staging? | Deploying now
Self-Review: Have you done an initial self-review of the code below on Github? | Yes
Spec Review: Have you reviewed the spec and ensured this meets requirements and/or matches design mockups? | N/A

anathomical committed:
Refactor OptionsQuery to take filter values

anathomical committed:
Add new options queries

anathomical committed:
Update SnapshotsController#options action to take filter params

This allows the back-end to update the front-end on which higher-specificity filters should be shown to the user based on broader filters that have already been applied.

anathomical committed:
Appease Rubocop linter.

The current configuration we have doesn't allow for open-ended ranges

anathomical committed:
Some more linter clean-up

The big one here is to declare our test class in a `let`

dandrabik commented on services/QuillLMS/app/controllers/snapshots_controller.rb:
```diff
@@ -37,10 +37,16 @@ def top_x
       schools: Snapshots::SchoolsOptionsQuery.run(current_user.id),
-      grades: GRADE_OPTIONS
+      grades: GRADE_OPTIONS,
+      teachers: Snapshots::TeachersOptionsQuery.run(current_user.id, school_ids, grades),
```
What would happen if one of the 3 BigQuery queries gives the "job not complete" error. I believe it would cause this endpoint to throw a 500. Does the frontend fetch them again in that case? With 3 jobs for an endpoint, seems like we might be upping that likelihood (although I don't have a good feel about how often that happens).

A few options that might be worth a discussion:
- Should we consider fetching these from the Postgres DB (and maybe limit to data to the end of the previous day). These aren't intense queries, and they might be more reliable coming from Postgres. 
- Should we consider making multiple endpoints? Right now, if you select a teacher_id, we only have to re-run the `classrooms` options, but we re-run all 5. If that's much more complex to implement, probably not worth the gain, but worth considering.

anathomical replied:
Actually, moving these queries to PostgreSQL is probably a good thing to do from a purely product perspective.  Right now if an admin is assigned a new school, their filter options won't show the new school until BigQuery updates that data.  So I think moving to PostgreSQL makes sense to do as part of this PR.

dandrabik commented on services/QuillLMS/app/queries/snapshots/period_query.rb:
```diff
@@ -4,7 +4,7 @@ module Snapshots
     attr_accessor :timeframe_start, :timeframe_end, :school_ids, :grades
 
-    def initialize(timeframe_start, timeframe_end, school_ids, grades = nil, options = {})
+    def initialize(timeframe_start, timeframe_end, school_ids, grades = nil, options: {})
```
I don't have strong feelings either way, but curious why this change? Changes to protocol for using the runner (the only use case) from `runner: runner` to `options: {runner: runner}`. NBD, just wondering if I'm missing something.

happythenewsad replied:
Same comment here, regarding keyword params

happythenewsad commented on services/QuillLMS/spec/queries/snapshots/options_query_spec.rb:
```diff
@@ -0,0 +1,68 @@
+      let(:results) {test_options_query.run(*query_args, options: {runner: runner}) }
+      # We want to ensure that a single admin is connected to all schools
+      # `admin` is already the user associated with the first school, but not the others
+      let(:additional_schools_admins) { schools[1..-1].map { |s| SchoolsAdmins.create(school: s, user: admin) } }
```
IMO `schools.filter { ... }.map ....`

would be more intelligible

happythenewsad commented on services/QuillLMS/app/queries/snapshots/options_query.rb:
```diff
@@ -2,10 +2,13 @@
+    attr_accessor :admin_id, :school_ids, :grades, :teacher_ids
 
-    def initialize(admin_id, options = {})
+    def initialize(admin_id, school_ids = nil, grades = nil, teacher_ids = nil, options: {})
```
Now that the method signature is getting longer, why not make `school_ids`, `grades`, and `teacher_ids` keyword params as well?

happythenewsad reviewed (approved):
Couple nits; looks good

dandrabik reviewed (commented):
Thanks for putting this together. I'm leaving this review as a "comment" since I think it overall looks good, and I'm not really requesting changes, but there are a few items that might be worth discussion. Those may incur large changes, so holding off on approving for that reason.

anathomical committed:
Use named params for queries since the list of args is getting long

anathomical committed:
WIP tweak options controller to use PostgreSQL rather than BigQuery

It's less brittle and also more up-to-date, which allows a newly-created admin to immediately access their reports

anathomical committed:
First pass at SnapshotController#options action in PostgreSQL

anathomical committed:
Refactor SnapshotController#options to use PostgreSQL

anathomical committed:
Remove BigQuery implementation of options queries

It makes more sense to pull them out of PostgreSQL for both performance and usability reasons

anathomical committed:
Add additional filters to PeriodQuery

To allow for filtering by classroom_id or teacher_id.  Also add spec coverage for root PeriodQuery behavior

anathomical committed:
Clean up trailing whitespace

anathomical committed:
Clean up specs to work with the updated param passing

anathomical committed:
Update cache logic to handle new filter params

anathomical committed:
Update specs for cache key generation based on new params

anathomical committed:
Refactor cache lib to meet linter max arg limits

anathomical committed:
Fix linting error, add a spec for cache keys with no extra filters

anathomical commented:
Re-requesting review on this PR as I've made extensive changes in light of moving the Options data query from BigQuery to PostgreSQL.

brendanshean commented on services/QuillLMS/app/controllers/snapshots_controller.rb:
```diff
@@ -4,7 +4,6 @@ class SnapshotsController < ApplicationController
   GRADE_OPTIONS = [
     {value: "Kindergarten", name: "Kindergarten"},
     {value: "1", name: "1st"},
-    {value: "2", name: "2nd"},
```
Just confirming no second grade?

anathomical replied:
Whoops!  Good catch!

brendanshean commented on services/QuillLMS/app/queries/snapshots/period_query.rb:
```diff
@@ -42,10 +46,22 @@ def school_ids_where_clause
 
-    def grade_where_clause
+    def grades_where_clause
       return "" if grades.nil? || grades.empty?
```
Would `return "" if grade.blank?` apply here?

brendanshean commented on services/QuillLMS/app/queries/snapshots/period_query.rb:
```diff
@@ -42,10 +46,22 @@ def school_ids_where_clause
     end
+
+    def teacher_ids_where_clause
+      return "" if teacher_ids.nil? || teacher_ids.empty?
```
Would `return "" if teacher_ids.blank?` apply here?

dandrabik replied:
consider `.blank?`

brendanshean commented on services/QuillLMS/app/queries/snapshots/period_query.rb:
```diff
@@ -42,10 +46,22 @@ def school_ids_where_clause
+    end
+
+    def classroom_ids_where_clause
+      return "" if classroom_ids.nil? || classroom_ids.empty?
```
Would `return "" if classroom_ids.blank?` apply here?

brendanshean commented on services/QuillLMS/app/lib/snapshots/cache_keys.rb:
```diff
@@ -2,16 +2,41 @@
     end
+
+    def self.grades_segment(grades)
+      return if grades.nil? || grades.empty?
```
Would `return "" if grades.blank?` apply here?

dandrabik replied:
`.blank?` is a short version of this line. I don't think you need the `grades || []` and this check. I believe that's redundant.

Also, you could consider not doing the nil check and the keys will all have the same sections, e.g. grades-teachers-classrooms.

brendanshean commented on services/QuillLMS/app/lib/snapshots/cache_keys.rb:
```diff
@@ -2,16 +2,41 @@
+    end
+
+    def self.teacher_ids_segment(teacher_ids)
+      return if teacher_ids.nil? || teacher_ids.empty?
```
Would `return "" if teacher_id.blank?` apply here?

brendanshean commented on services/QuillLMS/app/lib/snapshots/cache_keys.rb:
```diff
@@ -2,16 +2,41 @@
+    end
+
+    def self.classroom_ids_segment(classroom_ids)
+      return if classroom_ids.nil? || classroom_ids.empty?
```
Would `return "" if classroom_ids.blank?` apply here?

brendanshean commented on services/QuillLMS/spec/queries/snapshots/period_query_spec.rb:
```diff
@@ -0,0 +1,85 @@
+  describe PeriodQuery do
+
+    let(:test_period_query) do
+      Class.new(described_class) do
```
Cool!  I like this pattern.

anathomical replied:
I originally just declared a new `class` in the test file, but Rubocop has a rule against that.  So this pattern is one I stole from the Rubocop documentation for that rule.  Good job, linter!

brendanshean commented on services/QuillLMS/spec/controllers/snapshots_controller_spec.rb:
I like the thorough testing in this file.

brendanshean reviewed (approved):
Nice work on pulling all of these filters/options together.

dandrabik commented on services/QuillLMS/app/controllers/snapshots_controller.rb:
```diff
@@ -39,11 +38,48 @@ def top_x
+    grades = option_params[:grades]
+
+    filtered_schools = school_options
+    filtered_schools = filtered_schools.where(id: school_ids) unless school_ids.nil?
```
You should consider making `filtered_schools` or `filtered_school_ids` its own method to make this method less complex.

dandrabik commented on services/QuillLMS/app/controllers/snapshots_controller.rb:
```diff
@@ -39,11 +38,48 @@ def top_x
+    teacher_ids = option_params[:teacher_ids]
+
+    filtered_teachers = teacher_options
+    filtered_teachers = filtered_teachers.where(id: teacher_ids) unless teacher_ids.nil?
```
You should consider making `filtered_teachers` or `filtered_teacher_ids` its own method to make this method less complex.

dandrabik replied:
take or leave, consider the positive framing of this, e.g. `if teacher_ids.present?`

dandrabik commented on services/QuillLMS/app/controllers/snapshots_controller.rb:
```diff
@@ -39,11 +38,48 @@ def top_x
+      .joins(:classrooms_i_teach)
+      .where(schools_users: {school_id: filtered_schools.pluck(:id)})
+
+    teachers = teachers.where(classrooms: {grade: grades}) unless grades.nil?
```
take or leave, consider the positive framing of this, e.g. `if grades.present?`

dandrabik commented on services/QuillLMS/app/lib/snapshots/cache_keys.rb:
```diff
@@ -2,16 +2,41 @@
+      "grades-#{(grades || []).map(&:to_s).sort.join('-')}"
+    end
+
+    def self.teacher_ids_segment(teacher_ids)
```
Consider a helper method since you do this logic 4 times in the file, e.g.

```ruby
def id_segment(name, ids)
  #
end

dandrabik commented on services/QuillLMS/app/workers/snapshots/cache_snapshot_count_worker.rb:
```diff
@@ -17,18 +17,16 @@ class CacheSnapshotCountWorker
-          school_ids: school_ids,
-          grades: grades
-        }
+        }.merge(filters)
```
The cache key logic separates out school_ids from additional filters, and this puts school_ids in filters. Might be better to follow the same pattern in both.

dandrabik reviewed (approved):
Small code cleanup comments, but overall looks good to me. Thanks for moving this along!

anathomical committed:
Re-add an accidentally removed grade level

anathomical committed:
Update CacheKeys lib to DRY it out

anathomical committed:
Refactor PeriodQuery for cleaner syntax around early returns

Use .blank? over .nil? || .empty?

anathomical committed:
Refactor workers to have a similar signature to cache key generation

Specifically, pull `school_ids` out of the `filters` because school_ids is a required param for all queries.

anathomical committed:
Refactor controller to simplify code by breaking it into more methods
